synfig-studio/src/synfigapp/canvasinterface.cpp

```diff
@@ -699,6 +699,45 @@ CanvasInterface::jump_to_prev_keyframe()
 	//catch(...) { synfig::warning("Unable to find prev keyframe"); }
 }
 
+static void update_layer_size(CanvasInterface& canvas_interface, Layer::Handle& layer, bool resize_image) {
+	int w, h;
+
+	w = layer->get_param("_width").get(int());
+	h = layer->get_param("_height").get(int());
+	const RendDesc& rend_desc = canvas_interface.get_canvas()->rend_desc();
+	if (w && h) {
+		Vector x, size = rend_desc.get_br()-rend_desc.get_tl();
+
+		// vector from top left of canvas to bottom right
+		if (resize_image) {
+			if(abs(size[0])<abs(size[1])) {// if canvas is tall and thin
+				x[0]=size[0];	// use full width
+				x[1]=size[0]/w*h; // and scale for height
+				if ((size[0]<0) ^ (size[1]<0)) x[1] = -x[1];
+			} else { // else canvas is short and fat (or maybe square)
+				x[1]=size[1];	// use full height
+				x[0]=size[1]/h*w; // and scale for width
+				if ((size[0]<0) ^ (size[1]<0)) x[0] = -x[0];
```

Comment:
I particulary don't like these single-line `if`, but it's a matter of taste anyway :)

Reply:
Done